src/scripts/syncBreaches.js

```diff
@@ -11,7 +11,83 @@
 
 import { req, formatDataClassesArray } from '../utils/hibp.js'
 import { getAllBreaches, upsertBreaches } from '../db/tables/breaches.js'
+import { mkdir, readdir } from "node:fs/promises";
+import { resolve as pathResolve } from "node:path";
+import { finished } from 'stream/promises';
+import fs from "fs";
+import { Readable } from 'stream';
+import aws from 'aws-sdk';
+import { Upload } from "@aws-sdk/lib-storage";
 
+// Get breaches logos and uploads to s3
+const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
+const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
+const region = process.env.AWS_REGION;
+const Bucket = process.env.S3_BUCKET;
+
+const s3 = new aws.S3({
+  region,
+  credentials: {
+    accessKeyId,
+    secretAccessKey,
+  },
+});
+
+async function uploadToS3(fileName, fileStream) {
+  const uploadParams = {
+    Bucket,
+    Key: fileName,
+    Body: fileStream
+  }
+  try {
+  await s3.upload(uploadParams).promise()
+  console.log('Successfully uploaded data to ' + Bucket + '/' + fileName)
+} catch (err) {
+  console.error(err, err.stack)
+}
+}
+
+export async function getBreachIcons(breaches) {
+  let logoMap;
+  async function fetchBreachIcons() {
+    const breachDomains = breaches
+      .map((breach) => breach.Domain)
+      .filter((breachDomain) => breachDomain.length > 0);
+
+    // make logofolder if it doesn't exist
+    const logoFolder = "public/logo_cache/";
+    if (!fs.existsSync(logoFolder)) await mkdir(logoFolder);
+
+    // read existing logos
+    const existingLogos = await readdir(logoFolder);
+    console.log({existingLogos})
+
+    // TODO: Batch to limit memory use?
+    const logoMapElems = (await Promise.all(
+      breachDomains.map(async (breachDomain) => {
+          const logoFilename = breachDomain.toLowerCase() + ".ico";
+          const logoPath = pathResolve(logoFolder, logoFilename);
+          if (existingLogos.includes(logoFilename)) {
+            console.log('skipping ', logoFilename)
+            return;
+          }
+          console.log(`fetching: , ${logoFilename}`)
+          const res = await fetch(
+            `https://icons.duckduckgo.com/ip3/${breachDomain}.ico`);
+          const fileStream = fs.createWriteStream(logoPath, { flags: 'wx' });
```

Comment:
Interesting, does this need an "x" bit? (Looks like the [default `flags` is already `w`](https://nodejs.org/api/fs.html#fscreatewritestreampath-options), but maybe just curious why the "x" is there). Or maybe I'm completely missing the point. I assumed it was "x" for executable, but looks like it might be something else. Per https://nodejs.org/api/fs.html#file-system-flags:
> - `'wx'`: Like `'w'` but fails if the path exists.

Unclear on what the word "fails" means here. For example, if this (or any other async `fetch()`, `finished()`, or `uploadToS3()` `throw`s an error, I'm not sure who/where is listening/catching. Would one bad logo fetch break the entire `.map()` loop and abort loading any other logos?